pennylane/kernels/postprocessing.py

```diff
@@ -50,39 +51,137 @@ def displace_matrix(K):
+
+    Comments:
+        This method is introduced in https://arxiv.org/abs/2103.16774
+    """
+    w, v = np.linalg.eigh(K)
+
+    if np.min(w) < 0:
+        # Transform spectrum: absolute value
+        w0 = np.abs(w)
+
+        return (v * w0) @ np.transpose(v)
+
+    return K
 
-#     Args:
-#         K (array[float]): Kernel matrix assumed to be symmetric
-#         solver (str, optional): Solver to be used by cvxpy. Defaults to CVXOPT.
 
-#     Returns:
-#         array[float]: closest positive semidefinite matrix in Frobenius norm.
-#     """
-#     wmin = np.min(np.linalg.eigvals(K))
+def closest_psd_matrix(K, fix_diagonal=True, solver=None, **kwargs):
+    """Return the closest positive semidefinite matrix to the given kernel matrix.
 
-#     if wmin >= 0:
-#         return K
+    This method has the advantage that it achieves the correct diagonal entries
+    (fix_diagonal=True) or keeps the eigenvectors intact (fix_diagonal=False).
 
-#     X = cp.Variable(K.shape, PSD=True)
-#     objective_fn = cp.norm(X - K, "fro")
-#     problem = cp.Problem(cp.Minimize(objective_fn))
+    Args:
+        K (array[float]): Kernel matrix assumed to be symmetric.
+        fix_diagonal (bool): Whether to fix the diagonal to 1.
+        solver (str, optional): Solver to be used by cvxpy. Defaults to CVXOPT.
+        kwargs (kwarg dict): Passed to cvxpy.Problem.solve()
 
-#     try:
-#         problem.solve(solver=solver)
-#     except Exception as e:
-#         problem.solve(verbose=True, solver=solver)
+    Returns:
+        array[float]: closest positive semidefinite matrix in Frobenius norm.
 
-#     return X.value
+    Comments:
+        Requires cvxpy and the used solver (default CVXOPT) to be installed.
+        fix_diagonal=False typically leads to problems in the SDP/solving it.
+            Use `threshold_matrix()` instead, as it analytically is equivalent.
+    """
+    if not fix_diagonal:
+        wmin = np.min(np.linalg.eigvalsh(K))
+        if wmin >= 0:
+            return K
+    try:
+        import cvxpy as cp
+
+        if solver is None:
+            solver = cp.CVXOPT
+    except ImportError:
+        # TODO: Make these proper warnings
+        print("CVXPY is required for this post-processing method.", end="")
+        if fix_diagonal:
+            print(" As you want to fix the diagonal, task is impossible. Returning input...")
+            return K
+        else:
+            print(" As you don't want to fix the diagonal, using threshold_matrix instead...")
+            return threshold_matrix(K)
+
+    X = cp.Variable(K.shape, PSD=True)
+    constraint = [cp.diag(X) == 1.0] if fix_diagonal else []
+    objective_fn = cp.norm(X - K, "fro")
+    problem = cp.Problem(cp.Minimize(objective_fn), constraint)
+
+    try:
+        problem.solve(solver=solver, **kwargs)
+    except Exception as e:
+        problem.solve(verbose=True, solver=solver, **kwargs)
```

Comment:
```suggestion
        problem.solve(solver=solver, verbose=True, **kwargs)
```

Reply:
Extra arguments should be sorted in after the initial ones, otherwise this reads weirdly

Reply:
If the verbose one fails we should catch it again and raise a new `RuntimeError` that contains the original exception